### PR TITLE
The algorithm of percentage based sampling malfunctioned

### LINKED ISSUE #260: PercentageBasedSampler only works in increments of 10%
The threshold calculation in org.springframework.cloud.sleuth.sampler.PercentageBasedSampler breaks for percentages with resolution lower than 10%:

`long threshold = Math.abs(Long.MAX_VALUE * (int) (this.configuration.getPercentage() * 100)); // drops fractional percentage.`

This is is fine for settings like
`spring.sleuth.sampler.percentage=0.1`
`spring.sleuth.sampler.percentage=0.2`
`spring.sleuth.sampler.percentage=0.3`

But always returns true for settings such as
`spring.sleuth.sampler.percentage=0.01`
`spring.sleuth.sampler.percentage=0.15`
`spring.sleuth.sampler.percentage=0.79`

### PR DESCRIPTION
**Pull request by marcingrzejszczak**
I've reused the algorithm from Zipkin Sampler to fix it

fixes #260

CC @adriancole @dsyer @spencergibb

### CODE REVIEW

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/sampler/PercentageBasedSampler.java [hunk 1]
```diff
@@ -33,13 +33,13 @@ public PercentageBasedSampler(SamplerProperties configuration) {
 
 	@Override
 	public boolean isSampled(Span currentSpan) {
-		long threshold = Math.abs(Long.MAX_VALUE * (int) (this.configuration.getPercentage() * 100)); // drops fractional percentage.
+		long threshold = (long) (Long.MAX_VALUE * this.configuration.getPercentage()); // drops fractional percentage.
```
Comment (spencergibb) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/sampler/PercentageBasedSampler.java]: Comment no longer applies?

Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/sampler/PercentageBasedSampler.java]: comment should be `// safe cast as percentage < 1` or similar

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/sampler/PercentageBasedSamplerTests.java [hunk 2]
```diff
@@ -49,19 +49,36 @@ public void should_pass_given_percent_of_samples() throws Exception {
...
 	}
 
 	@Before
 	public void setupSpan() {
-		this.span = Span.builder().traceId(RANDOM.nextLong()).build();
+		this.span = newSpan();
+	}
+
+	Span newSpan() {
+		return Span.builder().traceId(RANDOM.nextLong()).build();
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/sampler/PercentageBasedSamplerTests.java]: yep. most importantly the sampler requires trace ids to be random, and.. the sample rate isn't stable until about 100,000 random ids pass through.

This means we might want to consider re-introducing a counting sampler, or making an option to make this one a counting one. If you have a low-traffic server, counting could be appropriate, given the warts.

@eirslett @kristofa I was thinking about this for brave, too.. wdyt?

**Comment by adriancole** [comment]
LGTM

**Pull request merged**